rfcs/0014-vm-cycle-limits/0014-vm-cycle-limits.md

```diff
@@ -0,0 +1,111 @@
+---
+Number: 0014
+Category: Standards Track
+Status: Proposal
+Author: Xuejie Xiao
+Organization: Nervos Foundation
+Created: 2019-01-04
+---
+
+# VM Cycle Limits
+
+## Introduction
+
+This RFC describes cycle limits used to regulate VM scripts.
+
+CKB VM is a flexible VM that is free to implement many control flow constructs, such as loops or branches. As a result, we will need to enforce certain rules in CKB VM to prevent malicious scripts, such as a script with infinite loops.
+
+We introduce a concept called `cycles`, each VM instruction or syscall will consume some amount of cycles. At consensus level, a scalar `max_block_cycles` field is defined so that the sum of cycles consumed by all scripts in a block cannot exceed this value. Otherwise, the block will be rejected. This way we can guarantee all scripts running in CKB VM will halt, or result in error state.
+
+## Consensus Change
+
+As mentioned above, a new scalar `max_block_cycles` field is added to chain spec as a consensus rule, it puts a hard limit on how many cycles a block's scripts can consume. No block can consume cycles larger than `max_block_cycles`.
+
+Note there's no limit on the cycles for an individual transaction or a script. As long as the whole block consumes cycles less than `max_block_cycles`, a transaction or a script in that block are free to consume how many cycles they want.
+
+## Cycle Measures
+
+Here we will specify the cycles needed by each CKB VM instructions or syscalls. Note right now in the RFC, we define hard rules for each instruction or syscall here, in future this might be moved into consensus rules so we can change them more easily.
+
+### Instruction Cycles
```

Comment:
How are these instruction cycles determined?

Reply:
This is roughly based on SiFive FU540's spec at [here](https://sifive.cdn.prismic.io/sifive%2F834354f0-08e6-423c-bf1f-0cb58ef14061_fu540-c000-v1.0.pdf), there're several basic principles:

* Branches should be more expensive than normal instructions
* Memory accesses should be more expensive than normal instructions, but since we are using 64-bit system, accessing 64-bit value should take less time than non 64-bit value
* Multiplication and divisions should be much more expensive than normal instructions